### DISCUSSION TITLE
How do I prevent node rearrangement?

### DISCUSSION THREAD

**Discussion by RapidTransit** [rehypejs/rehype#177] [Q&A] [answered]
So I'm using a SaaS app that uses a proprietary template language, one of the problems is I can't reuse fragments like say a breadcrumb template, like you could in other templating languages.

So I'm using Pug to transpile but even it's pretty option doesn't clean up the markup enough.

So I use I'm using rehype to clean up some white space in my html templates the problem is, rehype is rearranging tags it thinks are invalid markup and I can't figure out how to stop it. I tried enabling error reporting and the reporter plugin doesn't come back with anything.

How do I get it to stop moving my nodes around?

**Comment by wooorm** [comment]
Can you please provide some more info?

Share the actual input you have. The actual output you get. The expected output you want. The code you use to get there. The dependencies you have?

**Comment by RapidTransit** [comment]
Thank you for the quick reply, I'll create a reproducer

**Reply by wooorm** [reply]
I see you sent me an invite, to a private repo?
Can you make an open reproduction instead?
If there are private details, can you make the reproduction as small as possible, and omit the private details?

This is open source: I’m fine with helping folks out. But this is also for other, future readers, who need to see the code too.
I don‘t want to look at private code. That would be a job that I’d want to get paid for.

**Reply by RapidTransit** [reply]
Here ya go https://github.com/RapidTransit/rehype-reproducer

Thanks again.

**Reply by wooorm** [reply]
OK. I see the code. What is the problem with that? What is the output you want? How is it different from the output you get?

**Reply by RapidTransit** [reply]
The head tag becomes empty and the markup in the head tag end up in the body

Before:
```
 <head>
    <ac:templateoverridearea id="Mapper">
      <script type="importmap">{"imports":{"##THEMEDIR##js/app.js": "##FILEPATHVERSIONED[##THEMEDIR##js/app.js]##"##IF[SessionAdCode=DEVELOPER]##,"https://localhost:8080/js/app.js": "##FILEPATHVERSIONED[##THEMEDIR##js/app.js]##"##ENDIF##}}</script>
    </ac:templateoverridearea>
    <script defer type="module" src="##FILEPATHVERSIONED[##THEMEDIR##js/app.js]##"></script>$$HEADTAGS$$
    <ac:templateoverridearea id="ScriptsAddonArea"></ac:templateoverridearea>
    <ac:templateoverridearea id="HeadTagsAddonArea"></ac:templateoverridearea>
    <link rel="stylesheet" href="##FILEPATHVERSIONED[##THEMEDIR##css/app.css]##">
  </head>
  <body itemscope="">
    <div class="u-expand">
```

After:
```
 <head></head>
  <body itemscope>
    <ac:templateoverridearea id="Mapper">
      <script type="importmap">{"imports":{"##THEMEDIR##js/app.js": "##FILEPATHVERSIONED[##THEMEDIR##js/app.js]##"##IF[SessionAdCode=DEVELOPER]##,"https://localhost:8080/js/app.js": "##FILEPATHVERSIONED[##THEMEDIR##js/app.js]##"##ENDIF##}}</script>
    </ac:templateoverridearea>
    <script defer type="module" src="##FILEPATHVERSIONED[##THEMEDIR##js/app.js]##"></script>$$HEADTAGS$$
    <ac:templateoverridearea id="ScriptsAddonArea"></ac:templateoverridearea>
```

**Reply by wooorm** [reply]
It looks like you are not dealing with HTML, but with XML?

What are these things? `<ac:templateoverridearea>`?

Are you passing that HTML straight to browsers?

**Reply by RapidTransit** [reply]
Those are part of the template language, its a mix of tags and keywords, almost like Thymeleaf

**Reply by wooorm** [reply]
This project only understands HTML. It does not understand varying template languages.

**Reply by RapidTransit** [reply]
If it doesn't understand a tag, why would it move the tags around?

**Reply by wooorm** [reply, marked as answer]
That’s how html works. It does understand those tags. You can try it in a browser. It’s a very strong specification, which allows all kinds of things. One such things is that many tags are optional! The head opening for example. If you start with a div (or any unknown element, that cannot go in head, such as these), it’ll close the head and open the body for you

**Reply by RapidTransit** [reply]
OK, so it is intended behavior, I'll dig around the source and see if I can write my own patch.

**Reply by RapidTransit** [reply]
@wooorm Looks like it is a thing with parse5, it looks like I have to switch using htmlparser2 and made a parser visitor similar to Antlr.... I think this is the very definition of yak shaving

**Reply by wooorm** [reply]
I don’t understand? The tools here follow specs. What is Antlr?

If you want to support your own custom language, you’re free to do so, but yeah, you’ll have a terrible time to create all the tools that deal with your language.

Go with specifications. 🤷‍♂️ Lots of tools understand HTML. Use the things that exist.
